lib/pacemaker/pcmk_sched_clone.c

```diff
@@ -519,6 +519,113 @@ append_parent_colocation(pe_resource_t * rsc, pe_resource_t * child, gboolean al
     }
 }
 
+/*!
+ * \internal
+ * \brief Swap two instances' node allocations.
+ *
+ * \param[in,out] rsc1  The instance to be swapped with \c rsc2.
+ * \param[in,out] rsc2  The instance to be swapped with \c rsc1.
+ */
+static void
+swap_allocations(pe_resource_t *rsc1, pe_resource_t *rsc2)
+{
+    GHashTable *h_tmp = NULL;
+    pe_node_t *n_tmp = NULL;
+    pe_node_t *alloc1 = rsc1->fns->location(rsc1, NULL, 0);
+    pe_node_t *alloc2 = rsc2->fns->location(rsc2, NULL, 0);
+
+    pe_rsc_trace(rsc1->parent,
+                 "Swapping allocations: %s (currently allocated to %s) and %s "
+                 "(currently allocated to %s)", rsc1->id,
+                 alloc1->details->uname, rsc2->id, alloc2->details->uname);
+
+    /* Swap weights */
+    h_tmp = rsc1->allowed_nodes;
+    rsc1->allowed_nodes = rsc2->allowed_nodes;
+    rsc2->allowed_nodes = h_tmp;
+
+    /* alloc1 gets free'd during assign_node(rsc1, alloc2) */
+    n_tmp = pe__copy_node(alloc1);
+    assign_node(rsc1, alloc2, TRUE);
+    assign_node(rsc2, n_tmp, TRUE);
+    free(n_tmp);
+
+    /* @TODO: Do any of these need to be copied? */
+    // GListPtr rsc_cons_lhs;      // List of pcmk__colocation_t*
+    // GListPtr rsc_cons;          // List of pcmk__colocation_t*
+    // GListPtr rsc_location;      // List of pe__location_t*
+    // GListPtr actions;           // List of pe_action_t*
+    // GListPtr rsc_tickets;       // List of rsc_ticket*
+    // pe_node_t *partial_migration_target;
+    // pe_node_t *partial_migration_source;
+    // GHashTable *known_on;       /* pe_node_t*   */
+    // GListPtr children;          /* pe_resource_t*   */
+    // GListPtr dangling_migrations;       /* pe_node_t*       */
+    // pe_resource_t *container;
+    // GListPtr fillers;
+}
+
+/*!
+ * \internal
+ * \brief Reallocate a list of instances to their current nodes.
+ *
+ * This function assumes that all instances have already been allocated
+ * once. If a currently running instance is allocated to a different
+ * node, and its running-on node has also been allocated an instance,
+ * then we swap the allocations to try to allow as many instances as
+ * possible to remain on their current nodes.
+ *
+ * \param[in,out] instances     The list of instances to reallocate.
+ */
+static void
+reallocate_to_running_on(GListPtr instances)
+{
+    bool changed = true;
+
+    while (changed) {
```

Comment:
I'm fairly sure this outer loop is no longer needed. It's an artifact of an earlier implementation. I'm going to test without it now and update the PR if the tests pass.

Reply:
Fixed in the latest push.